statsd/sender.go

```diff
@@ -90,16 +96,30 @@ func (s *sender) flushTelemetryMetrics() SenderMetrics {
 func (s *sender) sendLoop() {
 	defer close(s.stop)
 	for {
+		// We occasionally unlock to signal to the flush function that
+		// we're done working on the current buffer.
+		s.flushLock.Lock()
 		select {
 		case buffer := <-s.queue:
 			s.write(buffer)
 		case <-s.stop:
+			s.flushLock.Unlock()
 			return
+		case <-time.After(receiveWaitTimeout):
+			break
 		}
+		s.flushLock.Unlock()
 	}
 }
 
 func (s *sender) flush() {
+	// At that point we know that the workers are paused (the statsd client
+	// will pause them before calling sender.flush()).
+	// But we have to wait for sendLoop to be done with
+	// it's current buffer, it any.
+	s.flushLock.Lock()
+	defer s.flushLock.Unlock()
```

Comment:
how about using a semaphore instead of using a lock, i.e.: `flush` sends a message on a dedicated channel on which `sendLoop` listens in its main loop. When `sendLoop` receives the message, it reads everything that's left in the buffer (guaranteed to complete since the workers are paused), signals back to `flush` that it's done, and goes back to its main loop. `flush` then returns.

I think that would allow to get rid of the `time.After` logic.

Reply:
+1